### PR TITLE
Project wide coverage/sanitization

### LINKED ISSUE #504: Make Coverage/Sanitizer build an option for the entire build directory
Currently we're generating special target for a number of targets that are compiled/linked with the coverage/sanitizer flags. This has some downsides:

- adding a new target (such as `hyriseSystemTest`) requires thinking about whether one wants coverage/sanitization for it. 
- If you decide something needs coverage/sanitization, you have to mess with cmake.
- Playground/Console don't have coverage/sanitization by default, but scenarios are thinkable where one would like to investigate something in the console with enabled sanitization.

I therefore propose to introduce cmake options `HYRISE_COVERAGE_ENABLED` and `HYRISE_SANITIZERS_ENABLED` which, if set to `TRUE` when calling `cmake` enable their respective option for ALL targets.
This would reduce cmake code, the number of targets that appear in the IDE menu and gets us rid of the `EXCLUDE_FROM_ALL` hacking we did a while back for the coverage/sanitization tests.

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fix #504 

Also, minimal CMake refactoring

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +16,23 @@ else()
     message(WARNING "You are using an unsupported compiler (${CMAKE_CXX_COMPILER_ID})! Compilation has only been tested with Clang and GCC.")
 endif()
 
-# Set default build type
+# Enable coverage if required
+option(HYRISE_COVERAGE "Set to ON to build Hyrise with enabled coverage checking. Default: FALSE" OFF)
```
Comment (mrks) [inline CMakeLists.txt]: on/off or true/false?

File: CMakeLists.txt [hunk 2]
```diff
@@ -15,7 +16,23 @@ else()
...
+if (${HYRISE_COVERAGE})
+    add_compile_options(-fprofile-arcs -ftest-coverage)
+    set(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} --coverage")
+endif()
+
+# Enable coverage if required
+option(HYRISE_SANITIZERS "Set to ON to build Hyrise with ASAN and UBSAN enabled. Defaul: FALSE" OFF)
```
Comment (mrks) [inline CMakeLists.txt]: see above, also: "Default"

File: Jenkinsfile [hunk 3]
```diff
@@ -21,10 +21,12 @@ node {
...
+        mkdir clang-debug-sanitizers && cd clang-debug-sanitizers && cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=clang-5.0 -DCMAKE_CXX_COMPILER=clang++-5.0 -DHYRISE_SANITIZERS=ON .. &\
+        mkdir clang-release-sanitizers && cd clang-release-sanitizers && cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang-5.0 -DCMAKE_CXX_COMPILER=clang++-5.0 -DHYRISE_SANITIZERS=ON .. &\
         mkdir clang-release && cd clang-release && cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang-5.0 -DCMAKE_CXX_COMPILER=clang++-5.0 .. &\
-        mkdir clang-release-no-numa && cd clang-release-no-numa && cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang-5.0 -DCMAKE_CXX_COMPILER=clang++-5.0 -DDISABLE_NUMA_SUPPORT=On .. &\
+        mkdir clang-release-sanitizers-no-numa && cd clang-release-sanitizers-no-numa && cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=clang-5.0 -DCMAKE_CXX_COMPILER=clang++-5.0 -DHYRISE_SANITIZERS=ON -DDISABLE_NUMA_SUPPORT=On .. &\
```
Comment (mrks) [inline Jenkinsfile]: With `DISABLE_NUMA_SUPPORT` we don't have a `HYRISE` prefix. Also: `ON` vs `On`.

**Comment by mrzzzrm** [comment]
cheers @mrks, but this is basically a PR-as-a-bookmark WIP thing right now ;)

**Comment by mrks** [comment]
I am in a train and forgot to download a movie first. :/

**Comment by mrzzzrm** [comment]
**Now** this is ready to be looked at.

**Pull request merged**